### PR TITLE
Retry storage access on internal errors

### PR DESCRIPTION
**Pull request by fische**
This revision adds a `retryBucket` struct that wraps the `bucket` interface in order to implement a retry logic on top of it. We go through this logic only on internal errors.

We are implementing this following an increase in the number of internal errors coming from reads on GCP buckets. The only suggested solution is to just retry when those errors occur: https://github.com/googleapis/google-cloud-go/issues/784

### CODE REVIEW

File: elan/rpc/retry.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+		err = f()
+	}
+	if err != nil && r > 0 {
+		return fmt.Errorf("%w (retries: %d)", err, r)
+	}
+	return err
+}
+
+func (b retryBucket) isErrorRetryable(err error) bool {
+	return gcerrors.Code(err) == gcerrors.Internal || status.Code(err) == codes.Internal
```
Comment (peterebden) [inline elan/rpc/retry.go:97]: I'd suggest that at least ResourceExhausted and DeadlineExceeded should also be retryable. Probably also Canceled and maybe Unknown too (that's debatable but may be better to retry pointlessly than misclassify an error)

File: elan/rpc/retry.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+// retryBucket implements a retry logic on top of all bucket methods that can
+// return an error. Those methods are retried only on internal errors that are
+// not caused by a context cancellation.
+type retryBucket struct {
+	bucket
+	retries int
+}
+
+func (b retryBucket) NewRangeReader(ctx context.Context, key string, offset, length int64, opts *blob.ReaderOptions) (io.ReadCloser, error) {
+	var (
```
Comment (peterebden) [inline elan/rpc/retry.go:23]: I feel like generics might provide a nicer way of sharing the implementation of these?

**Review by peterebden — COMMENTED** [review]
Looks good but this doesn't handle the case where mettle workers go directly to storage (probably that wants a retryable wrapper around the 'clients' in this package too).

It also won't retry if you get an initial reader or writer successfully but it fails later. Do we know if that happens or if it's generally upfront that it fails?

**Comment by peterebden** [comment]
You might also want to rebase on master - #181 eventually got a green build out of golangci.

File: elan/rpc/retry.go (hunk 1 above)
Comment (fische) [reply elan/rpc/retry.go:97]: Yeah good point about `ResourceExhausted` and `Unknown`. Regarding `DeadlineExceeded` and `Canceled` I'll still check if this is coming from our context, in which case there's no point in retrying.

**Comment by fische** [comment]
> Looks good but this doesn't handle the case where mettle workers go directly to storage (probably that wants a retryable wrapper around the 'clients' in this package too).

Do you mean when we use the `elanClient` instead of the `remoteClient` implementation? If that's what you mean, it's already using this wrapper as `New` calls `createServer`.

> It also won't retry if you get an initial reader or writer successfully but it fails later. Do we know if that happens or if it's generally upfront that it fails?

That's a very good point. There's no indication that this is only happening when initialising the reader/writers, so I'll add retries for that as well.

**Comment by fische** [comment]
Abandoning this PR, as we can just use the retry logic from the GCS client, we just need to update the storage package for that.

**Pull request closed without merging**